Discuss combining and retiring logs. (#10)
ietf-wg-keytrans/draft-arch: draft-ietf-keytrans-architecture.md

Transparency Log to the Manager. Rejected requests are blocked." }

# Combining Logs

There are many cases where it makes sense to operate multiple cooperating log
instances, for example:

> - A service provider may decide that it's prudent to rotate its cryptographic

Reviewer: ```suggestion
- A service provider MAY decide that it's prudent to rotate its cryptographic
```
is this MAY, or only a suggestion?

  keys, or migrate to a new deployment mode. They can do this by creating a new
  log instance with new cryptographic keys, operating under a new
  deployment mode if desired, and migrating their data from the old log to the
  new log while users are able to query both.
- A service provider may choose to operate multiple logs to improve their
  ability to scale or provide higher availability.
- A federated system may allow each participant in the federation to operate

>   their own log for their own users.

Reviewer: do you envision sharding to be used, by year for instance? I have the impression it provides a slightly more concrete example.

Client implementations should generally be prepared to interact with multiple

> logs simultaneously. In particular, clients SHOULD namespace any configuration

Reviewer 2: What happens if conflicts exist? Why not MUST?

Author: "SHOULD" because I think supporting multiple logs should be the default decision, unless an implementer really really knows they're only ever going to talk to one log at a time

or state related to a particular log, such that information related to different
logs do not conflict.

When multiple logs are used, all users in the system MUST have a consistent

> policy for executing Search, Update, and Monitor queries against the logs in a

Reviewer 2: I wonder if an example policy might make this clearer... It reads as " the service provider should have a policy of not being duplicitous"...but if that is detectable, what does this MUST really unlock?

Author: This "MUST" doesn't affect the service provider, it's saying all of the users need to behave in a consistent way. If they don't, it gets very hard to reason about whether security guarantees are met

Reviewer 2: So this is a policy suggestion for "clients" ? maybe client is a better word than user here.... saying "must be consistent" without further clarity remains triggering for me, but your answer to makes sense to me.

way that maintains the high-level security guarantees of KT:

- If all logs behave honestly, then users observe a globally-consistent view of
  the data associated with each key.
- If any log behaves dishonestly such that the prior guarantee is not met (some
  users observe data associated with a key that others do not), this will be
  detected either immediately or in a timely manner by background monitoring.

## Gradual Migration

In the case of gradually migrating from an old log to a new one, this policy may
look like:

1. Search queries should be executed against the old log first, and then against
   the new log only if the most recent version of a key in the old log is a

>    tombstone.

Reviewer 3: "Tombstone" is used here for the first time without further definition. I think this is a concept that was first introduced in SEEMless and requires introduction.

2. Update queries should only be executed against the new log, adding a
   tombstone entry to the old log if one hasn't been already created.
3. Both logs should be monitored as they would be if they were run individually.
   Once the migration has completed and the old log has stopped accepting

>    changes, the old log SHOULD stay operational long enough for all users to

Reviewer 2: Again what happens when this doesn't happen, why not MUST?

   complete their monitoring of it (keeping in mind that some users may be
   offline for a significant amount of time).

Placing a tombstone entry for each key in the old log gives users a clear
indication as to which log contains the most recent version of a key and
prevents them from incorrectly accepting a stale version if the new log rejects
a search query.

## Immediate Migration

In the event of a key compromise, the service provider may instead choose to
stop adding new entries to a log immediately and provide a new log that is
pre-populated with the most recent versions of all keys. In this case, the
policy may look like:

> 1. Search queries should always be executed against the new log.

Reviewer: ```suggestion
1. Search queries must be executed against the new log.
```
semantic, should always is a must I would say

   own are properly represented in the new log. From then on, users will monitor
   only the new log.

The final tree size and root hash of the prior log must be distributed to users
in a way that guarantees all users have a globally-consistent view. This can be
done either by storing them in a well-known key of the new log, or with the
application's code distribution mechanism.

## Federation

In a federated application, many servers that are owned and operated by
different entities will cooperate to provide a single end-to-end encrypted
communication service. Each entity in a federated system provides its own
infrastructure (in particular, a transparency log) to serve the users that rely
on it. Given this, there must be a consistent policy for directing KT requests
to the correct transparency log. Typically in such a system, the end-user
identity directly specifies which entity requests should be directed to. For
example, with an email end-user identity like `alice@example.com`, the
controlling entity is `example.com`.

A controlling entity like `example.com` may act as an anonymizing proxy for its
users when querying transparency logs run by other entities, but should not

> attempt to 'mirror' or combine other transparency logs with its own.

Reviewer: for an HTTP request, this would be `no-store`/`private`?

Author: I think we are looking at something more like OHAI. So the server in the federation that I rely on can anonymize my requests to other servers, but doesn't end up observing a bunch of private user data.
